### PR TITLE
more basic functionality added to Geochemistry module

### PR DESCRIPTION
**Pull request by WilkAndy**

### CODE REVIEW

**Comment by WilkAndy** [comment]
Hey @cpgr - i'd appreciate a review on this.  Most of it is quite simple and shouldn't be difficult to review.  The redox stuff is the most complicated in terms of geochemistry - i'm pretty convinced i've got all the swaps and stuff correct because i've tested it quite thoroughly.

Hey @loganharbour - are you sick of these geochem PRs yet?  If so, i'm sure i can hassle someone else - don't feel bad about saying "too busy".

**Comment by cpgr** [comment]
I'll take a look tonight

**Comment by WilkAndy** [comment]
Failure is ` TIMEOUT navier_stokes/test:ins/boussinesq.jac_stab` unrelated to this PR

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    _dhatilde(0.0),
+    _dhbtilde(0.0),
+    _dhctilde(0.0),
+    _dhdtilde(0.0)
+{
+}
+
+void
+GeochemistryActivityCoefficients::setInternalParameters(
+    Real /*temperature*/,
```
Comment (cpgr) [inline modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:38]: Why is temperature not needed?

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  _dhatilde = 1.45397;
+  _dhbtilde = 0.022357;
+  _dhctilde = 0.0093804;
+  _dhdtilde = -0.0005262;
+  // neutral species Debye-Huckel at 25degC
+  _dha = 0.1127;
+  _dhb = -0.01049;
+  _dhc = 0.001545;
+  _dhd = 0.0;
+}
```
Comment (cpgr) [inline modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:66]: Oh, I see. Waiting until you can interpolate them.

File: modules/geochemistry/src/utils/PertinentGeochemicalSystem.C [hunk 3]
```diff
@@ -477,6 +481,79 @@ PertinentGeochemicalSystem::createModel()
...
+          for (unsigned basis_i = 0; basis_i < num_cols; ++basis_i)
+            stoi[basis_i] += _model.eqm_stoichiometry(free_el_index, basis_i);
+          // now the reation is e- = nuw_i * basis_i - beta/alpha * (-redox + nu_i * basis_i)
+          redox_stoi.insert(redox_stoi.end(), stoi.begin(), stoi.end());
+
+          // record the equilibrium constants
+          for (unsigned temp = 0; temp < num_temperatures; ++temp)
+            redox_log10K.push_back((-beta / alpha) * rs.equilibrium_const[temp] +
+                                   _model.eqm_log10K(free_el_index, temp));
+        }
```
Comment (cpgr) [inline modules/geochemistry/src/utils/PertinentGeochemicalSystem.C:543]: Thanks for the comments explaining the reactions. Makes it much easier to follow!

**Comment by loganharbour** [comment]
Feel free to keep pinging me for reviews. I'll defer if need be, but I'm definitely good now.

I'll wait until you address the comments from Chris.

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "PertinentGeochemicalSystem.h"
+#include "GeochemistryIonicStrength.h"
+
+/**
+ * Computes activity coefficients for non-minerals and non-gases (since these are meaningless). Also
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h:16]: yep

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  /**
+   * Set internal parameters, such as the ionic strength and Debye-Huckel parameters, prior to
+   * computing activity coefficients and activity of water.  Ensure the ionic strength calculator
+   * (eg, maxIonicStrength) is set appropriately before calling this function.
+   * @param temperature the temperature in degC
+   * @param mgd the Model Geochemical database
+   * @param basis_species_molality Molalities of the basis species in mgd
+   * @param eqm_species_molality Molalities of the equilibrium species in mgd
+   * @param kin_species_molality Molalities of the kinetic species
+   * @return the ionic strength of the aqueous solution
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h:48]: thanks.  Deleted

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  };
+
+  /**
+   * @param method Method used by this class to compute activity coefficients and activity of water
+   * @param max_ionic_strength maximum value of ionic strength (used if method=DEBYE_HUCKEL)
+   * @param max_stoichiometric_ionic_strength maximum value of stoichiometric ionic strength (used
+   * if method=DEBYE_HUCKEL)
+   * @param use_only_basis_molality If true, use only the basis molalities in the ionic strength and
+   * stoichiometric ionic strength calculations (used if method=DEBYE_HUCKEL)
+   */
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h:35]: Oh thanks.  Deleted

File: modules/geochemistry/include/utils/GeochemistryFormattedOutput.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#include "MooseTypes.h"
+#include "DenseMatrix.h"
+#include <libmesh/dense_vector.h>
+
+namespace GeochemistryFormattedOutput
+{
+/**
+ * Returns a nicely formatted string corresponding to the reaction defined by the given row of the
+ * stoichiometric matrix. For instance, if stoi(row, :) = (1, 2, 1E-5, -0.333333333333) and names =
+ * ["H2O", "H+", "HCO3-", "O2(aq)"] then this function will return "H2O + 2*H+ - 0.333O2(aq)" if
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryFormattedOutput.h:21]: Ha, yes, ooops.  Changed.

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+GeochemistryActivityCoefficients::getIonicStrength() const
+{
+  return _ionic_strength;
+}
+
+Real
+GeochemistryActivityCoefficients::getStoichiometricIonicStrength() const
+{
+  return _stoichiometric_ionic_strength;
+}
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:180]: If we ever implement another sort of activity model, what i had imagined is that we'd add more stuff to
```
  /// Method used by this class to compute activity coefficients and activity of water                                     
  enum class ActivityCoefficientMethodEnum
  {
    DEBYE_HUCKEL
  };
```
and more code to the .C file.   But i think you're right - i'll make the change.

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C (hunk 1 above)
Comment (WilkAndy) [reply modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:38]: Yep, waiting on that PR of yours to be merged!

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C (hunk 2 above)
Comment (WilkAndy) [reply modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:66]: yep :-)

File: modules/geochemistry/src/utils/PertinentGeochemicalSystem.C (hunk 3 above)
Comment (WilkAndy) [reply modules/geochemistry/src/utils/PertinentGeochemicalSystem.C:543]: Yea, this stuff kills my mind sometimes!

File: modules/geochemistry/include/utils/GeochemistryFormattedOutput.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * in the returned string
+ * @param precision Maximum number of digits written for each stoichiometric coefficient in the
+ * returned string
+ */
+std::string reaction(const DenseMatrix<Real> & stoi,
+                     unsigned row,
+                     const std::vector<std::string> & names,
+                     Real stoi_tol,
+                     int precision);
+}
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryFormattedOutput.h:36]: Yea, ok, good idea.  Done

File: modules/geochemistry/src/utils/GeochemistrySpeciesSwapper.C [hunk 10]
```diff
@@ -167,6 +169,29 @@ GeochemistrySpeciesSwapper::alterMGD(ModelGeochemicalDatabase & mgd,
...
+  if (redox_lhs_going_to_basis)
+  {
+    // need to make the redox_lhs equal to the species that is being taken from the basis
+    mgd.redox_lhs = basis_name;
+    for (unsigned red = 0; red < num_redox; ++red)
+    {
+      const Real alpha = mgd.eqm_stoichiometry(
+          eqm_index_to_insert, basis_index_to_replace); // must be nonzero due to valid swap
+      const Real alpha_r = mgd.redox_stoichiometry(red, basis_index_to_replace);
+      const Real coef = 1.0 / (alpha - alpha_r);
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/src/utils/GeochemistrySpeciesSwapper.C:184]: Yea, alpha!=alpha_r if previous coding is correct.  But i'll include a mooseAssert

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+   * @return the activity of water
+   */
+  Real waterActivity() const;
+
+  /**
+   * Compute the activity coefficients and store them in basis_activity_coef and eqm_activity_coef
+   * Note:
+   * - you will probably want to call setInternalParameters prior to calling this method
+   * - the activity coefficient for water (basis species = 0) is not computed since it is
+   * meaningless: get getWaterActivity() instead
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficients.h:68]: thanks, changed

File: modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+  _dhb = -0.01049;
+  _dhc = 0.001545;
+  _dhd = 0.0;
+}
+
+std::vector<Real>
+GeochemistryActivityCoefficients::getDebyeHuckel() const
+{
+  return {_dhA, _dhB, _dhBdot, _dhatilde, _dhbtilde, _dhctilde, _dhdtilde, _dha, _dhb, _dhc, _dhd};
+}
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/src/utils/GeochemistryActivityCoefficients.C:72]: Done

**Comment by WilkAndy** [comment]
Thanks for all those suggestions, @cpgr.  I've implemented them all.   Now over to you @loganharbour

**Comment by cpgr** [comment]
Looks great Andy!

**Comment by WilkAndy** [comment]
Bugging @loganharbour

**Comment by loganharbour** [comment]
Sorry, Andy. This'll come either this evening or tomorrow morning.

**Comment by WilkAndy** [comment]
Cool, thanks @loganharbour - looking forward to it!

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Computes activity coefficients for non-minerals and non-gases (since these species do not have
+ * activity coefficients). Also computes the activity of water.  Uses a Debye-Huckel model
+ */
+class GeochemistryActivityCoefficientsDebyeHuckel : public GeochemistryActivityCoefficients
+{
+public:
+  /// Method used by this class to compute activity coefficients and activity of water
+  enum class ActivityCoefficientMethodEnum
```
Comment (loganharbour) [inline modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h:53]: I believe this should be in the base class as a virtual. The base class will then (or some helper) hold an enum of all the valid types, and the children classes will override with their type of these.

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h (hunk 13 above)
Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h:53]: Great, thanks @loganharbour .  I got up especially early to address your change requests, and hopefully we can get this merged today.  I'll work on this request shortly....

Comment (loganharbour) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h:53]: 👍 I'll be around for a while

Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h:53]: OK, i cannot forward-declare an enum and define it in a derived class.  So I think this suggestion is not going to work.  Also, I don't want derived classes such as `GeochemistryActivityCoefficientsDebyeHuckel` knowing about other activity-coefficient methods that will be used in other derived classes (i know there are no other derived classes at the moment).

Instead, I'm simply going to remove this enum entirely.  It was premature to put it in - i thought that someone might add to it one day to implement different DebyeHuckel-style activity calculators inside `GeochemistryActivityCoefficientsDebyeHuckel`   But i'll let the do that in the future.

Comment (loganharbour) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h:53]: Okay, that's fine. The way this is typically done is that _some_ object needs to know about all of the possible choices. Typically, we'll declare that in the parent object or using some registry system.

Comment (WilkAndy) [reply modules/geochemistry/include/utils/GeochemistryActivityCoefficientsDebyeHuckel.h]: OK, cool.  Let's watch those green ticks appear....

**Comment by WilkAndy** [comment]
Cool - all green @loganharbour

**Review by loganharbour — APPROVED** [review]
👍 

I think that at some point, because this module is mostly a set of utilities with a wrapper, we should organize the utils into separate folders too.

**Pull request merged**